### ISSUE TITLE
[HOLD for payment 2023-09-12] [HOLD for payment 2023-09-11] [$1000]  [Distance] - App crashes when opening IOU thread and IOU shows a red dot

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#26585]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Issue found when executing PR** https://github.com/Expensify/App/pull/25707
## Action Performed:
1. Launch New Expensify app
2. Go to workspace chat
3. Go to + > Request money > Distance
4. Create a distance request
5. Right after the request is created (when it shows $0.00), click on the IOU to open IOU thread and wait for a while
6. Relaunch New Expensify app
7. Go to the created distance IOU

## Expected Result:
In Step 5, app does not crash
In Step 7. the distance IOU does not have a red dot

## Actual Result:
In Step 5, the app crashes
In Step 7. the distance IOU shows a red dot

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.62-1

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/81deaafc-3c8b-43a7-b61f-36fa533317f3

[aaaaaaaa.txt](https://github.com/Expensify/App/files/12504863/aaaaaaaa.txt)

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team and also @situchan for [a separate issue](https://expensify.slack.com/archives/C049HHMV9SM/p1693763384079999) solved by the PR for this issue

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014333c9073e9221c1</li>
        <li>Upwork Job ID: 1698167577340440576</li>
        <li>Last Price Increase: 2023-09-03</li>
<li>Automatic offers: </li>
<ul>
<li>pradeepmdk | Contributor | 26478625</li>
<li>Pujan92 | Contributor | 26480728</li>
</ul></li>
    </ul>
</details>

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
App crashes when opening IOU thread and IOU shows a red dot

### What is the root cause of that problem?
**in the latest main red dot  is already fixed.**
URI is passing as number because  because local import should be passed with without uri key.
<img width="1263" alt="Screenshot 2023-09-03 at 5 14 12 AM" src="https://github.com/Expensify/App/assets/16595846/e882f9ff-ad2e-4eb5-b941-283192407fcb">

why because it fetching local image means here file name is empty when the `CreateDistanceRequest` API called
<img width="957" alt="Screenshot 2023-09-03 at 6 21 03 AM" src="https://github.com/Expensify/App/assets/16595846/ce12188e-ecf4-4ccd-9e78-fc6c2bc3b652">

but in the API we have `receiptFilename`


### What changes do you think we should make in order to solve the problem?

form BC add fileName in the api

or
https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/components/ReportActionItem/MoneyRequestPreview.js#L226
here we need add  
`props.transaction.filename || props.transaction.receiptFilename `

### Alternative Solution

here in ReportActionItemImage uri will doesn't support local assets file path. 
```
 return (
        <Image
            source={{uri: image}}
            style={[styles.w100, styles.h100]}
        />
    );
``` 
so we need 'ReportActionItemImage' one more props if we are trying to load assets image means 
```
 return (
        <Image
            source={assets ?  assets : {uri: image}}
            style={[styles.w100, styles.h100]}
        />
    );
``` 
and this function should return `getThumbnailAndImageURIs` {thumbnail: null, image: null , assets: localImagepath};


I already added commented another PR https://github.com/Expensify/App/issues/25841

please consider this as well

**Comment by hayata-suenaga** [comment]
It's weekend and we need to fix this by Monday so I gonna make this external to get immediate proposals

**Comment by pradeepmdk** [comment]
IOU shows a red dot already fixed. here https://github.com/Expensify/App/issues/26567

**Comment by hayata-suenaga** [comment]
@pradeepmdk I appreciate your immediate solution and I feel like your proposal is correct

I just need more help understanding your solution.

could you clarify that is the exact issue and how we can solve it?

**Comment by pradeepmdk** [comment]
@hayata-suenaga  `CreateDistanceRequest` api response filename is missing 
so here when filename is not there means it will return local file https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/libs/ReceiptUtils.js#L40

https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/components/ReportActionItem/ReportActionItemImage.js#L75 here we are passing local file in uri so crash happened here 


we need fix this on backend by adding the `filename`
form UI means 
now backend return have the `receiptFilename`  so we need add  like this 

`props.transaction.filename || props.transaction.receiptFilename `
https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/components/ReportActionItem/MoneyRequestPreview.js#L226

**Comment by hayata-suenaga** [comment]
@pradeepmdk 

I couldn't reproduce the issue on the web. could you check if the issue is reproducible on the web, too? 🙇 🙇 🙇

**Comment by pradeepmdk** [comment]
@hayata-suenaga This happens only on native devices only. because native ios and android expecting the URI is as a string.

**Comment by hayata-suenaga** [comment]
@pradeepmdk sorry for the late response I'm juggling several issues right now 😓 

My question is, the preview displays the static map when it's loaded right? I was wondering where the url is coming from if the filename is not present on API responses?

My question might not make sense but I'm just trying my best to quickly understand the context 🙇

**Comment by pradeepmdk** [comment]
<img width="1301" alt="Screenshot 2023-09-03 at 10 37 43 AM" src="https://github.com/Expensify/App/assets/16595846/5ac8d974-cafd-4b04-8717-1c5c4878ff12">
@hayata-suenaga  in web here map showing as an image only. so after creating `CreateDistanceRequest` its coming to workspace here only at the time filename is not there so in native render  try load local image so its breaking

**Comment by b4s36t4** [comment]
@pradeepmdk does this issue happens when you use staging server? you can turn that on preferences. Can you update about that?

**Comment by pradeepmdk** [comment]
@b4s36t4 no it's not related staging server.  main issue is we are passing local image in wrong source. we need avoid to load local image we need filename in the `CreateDistanceRequest` api or else in the ui we need use `receiptFilename`

**Comment by b4s36t4** [comment]
@pradeepmdk I don't understand the concept of local image at all, for every image we have authentication enabled which requires to load the image everytime from network itself. Even if there's cache it's handled by react-native itself, so I'm a bit confused here. Can you explain this Hope I'm not asking wrong questions or so.

**Comment by b4s36t4** [comment]
whenever we pass source as an object to react-native image it is meant that image is loading from a server not from the local. If the image is local we need to provide the source as following way

```
source={require("../image.png")}
```

**Comment by pradeepmdk** [comment]
@b4s36t4  we using this function https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/libs/ReceiptUtils.js#L40  if filename not found we are returning local image so at the time are passing as   `source={uri: image}`

**Comment by pradeepmdk** [comment]
@hayata-suenaga  in the front side as mentioned in the proposal we need to fix this 
https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/components/ReportActionItem/ReportActionItemImage.js#L75 source so that anywhere image is not breaking  
for this case as well until static map load we can see the local image 
![Screenshot 2023-09-03 at 11 09 50 AM](https://github.com/Expensify/App/assets/16595846/1e313c9b-e7a7-4d7f-98b6-b64d51194683)

currently its working perfectly in the web only native its breaking

**Comment by hayata-suenaga** [comment]
@pradeepmdk

> image is load based on the filename only at the time filename is missing so try to use localimage but in our component we are loading local image in uri

are you taking about optimistic data that we use until real data is loaded?

**Comment by pradeepmdk** [comment]
@hayata-suenaga that is also breaking here 
https://github.com/Expensify/App/blob/d0b277206ff8b80f8b447ecb49437cec0446f2f8/src/components/ReportActionItem/ReportPreview.js#L182  here they are extracting the filename but `filename` not there so going as a undefined 
![Screenshot 2023-09-03 at 12 48 32 PM](https://github.com/Expensify/App/assets/16595846/11b545a9-4359-43ad-9c9d-24f93e64bef8)
and here line is breaking 
https://github.com/Expensify/App/blob/d0b277206ff8b80f8b447ecb49437cec0446f2f8/src/libs/fileDownload/FileUtils.js#L104
here solution is 
```javascript
ReceiptUtils.getThumbnailAndImageURIs(receipt.source, filename || '')
```

**Comment by pradeepmdk** [comment]
so after fixing  this above one  same issue here as well 
![Screenshot 2023-09-03 at 12 52 42 PM](https://github.com/Expensify/App/assets/16595846/bf84a9cc-fdb7-4f6f-9fcc-f55f845d4b7a)

**Comment by hayata-suenaga** [comment]
I'm reviewing your proposal right now @pradeepmdk

**Comment by pradeepmdk** [comment]
@hayata-suenaga If we fix this https://github.com/Expensify/App/issues/26585#issuecomment-1704012836 it will solve this perfectly for this crash issue, until the server image we can show the local image.

**Comment by hayata-suenaga** [comment]
@pradeepmdk thank you for that

so that is going to solve the issue for both receipt and map reviews right?

I gonna assign you right now so that you can start working on the PR. Might ask follow up questions once the PR is up

**Comment by b4s36t4** [comment]
I'm not able to re-produce this. @pradeepmdk @hayata-suenaga are you guys able to?

**Comment by pradeepmdk** [comment]
>so that is going to solve the issue for both receipt and map reviews right?

yes

**Comment by b4s36t4** [comment]
You're on latest pull?

**Comment by b4s36t4** [comment]
Even after click right after the IOU created, it's loading the image for me. Any device specific android or ios?

**Comment by b4s36t4** [comment]
in iOS it seems to be working.

**Comment by pradeepmdk** [comment]
android mainly cause this due to strict type validation

**Comment by b4s36t4** [comment]
Okay got it, but does your solution considers about iOS? Hope you're aware of the conditions.

**Comment by pradeepmdk** [comment]
@parasharrajat PR is ready

**Comment by Pujan92** [comment]
I think this issue is more of a backend inconsistency in the fields.

Pusher event and Create distance api sending field as `receiptFilename `(eg. `"receiptFilename":"w_143ae70302c853c99a25c03c88f4ca42848ff717.png"`)

<img width="1440" alt="Screenshot 2023-09-04 at 00 54 18" src="https://github.com/Expensify/App/assets/14358475/3ea3495e-990b-4193-b24b-eb9cbb8219a1">


Whereas OpenReport responding the transactions with the field `fileName`

<img width="632" alt="Screenshot 2023-09-04 at 00 53 57" src="https://github.com/Expensify/App/assets/14358475/e991920c-79db-4e38-9475-56bacd994603">

Maybe everywhere it should be considered as as a `fileName`

**Comment by pradeepmdk** [comment]
@Pujan92 yes. I already addressed this in the proposal not only here scan receipt API also has the same issue filename is not there.

**Comment by Pujan92** [comment]
Passing the empty string for the `filename` only prevents a crash but it doesn't seem the correct root cause here. 
Backend change will also render the correct image whenever the pusher event receives a new transaction(bcoz if we received the `filename` then it can be used correctly with the current logic)

**Comment by pradeepmdk** [comment]
[App/src/libs/ReceiptUtils.js](https://github.com/Expensify/App/blob/e088539f32561dca5e0490300bd9e5684cd54d04/src/libs/ReceiptUtils.js#L40) if filename empty we are returning local image.

**Comment by hayata-suenaga** [comment]
@Pujan92 @pradeepmdk so we're not able to display receipt or map in any screen right now?

**Comment by pradeepmdk** [comment]
@hayata-suenaga not like that. when `OpenReport` API called means filename is there so image will show.

**Comment by hayata-suenaga** [comment]
@pradeepmdk 

I really appreciate your prompt responses, but could you aim for more clear explanations when responding (for example, using complete sentences)?

So, what I'm confused about is; which API command lacks the filename? and which API command properly returns the filename?

Why the font end change alone is not enough? When will we have an issue if we don't have the backend change?

**Comment by Pujan92** [comment]
@hayata-suenaga If we only provide an empty filename for `ReportPreview.js` [here](https://github.com/Expensify/App/pull/26611/files#diff-90dbf92cac79a4d1c6199362686c19700554305c62fbbeadcaa509c3abe7e5a9R182), issue still persists for `MoneyRequestPreview`

1) App crashes when admin receives a new distance request pusher event due to the `filename` not available

https://github.com/Expensify/App/assets/14358475/689354e4-fdc3-4861-a290-65847866aa75

2) If we pass empty for filename in `MoneyRequestPreview` too, it won't render the correct image after the pusher event

https://github.com/Expensify/App/assets/14358475/68fc2568-732d-4243-81cb-b732184a5660

So, according to me first we need to make changes in the backend by responding with the `filename` instead of `receiptfilename` in the apis and pusher events.

**Comment by pradeepmdk** [comment]
@Pujan92 I agree with this but here we are fixing the crash issue on Android source is not passing correctly https://github.com/Expensify/App/issues/26585#issuecomment-1703976018. 

@hayata-suenaga as per @Pujan92 videos is correct because we are not receiving the file name means we are showing the local image until we get the real image. so that in `CreateDistanceRequest ` and `pusher` send the filename means we will load the real image or else we need to wait until load the `openreport` api.

**Comment by pradeepmdk** [comment]
>  which API command lacks the filename
pusher , CreateDistanceRequest, RequestMoney

>which API command properly returns the filename? 
openreport

>Why the font end change alone is not enough? When will we have an issue if we don't have the backend change? 
we have the filename only `openreport` api now. so until then we are showing the local image

**Comment by hayata-suenaga** [comment]
okay that makes things more clear now

@Pujan92 I also gonna assign to this issue so that we don't forget to handle partial payment for you for your continuous help with this

**Comment by hayata-suenaga** [comment]
@JmillsExpensify 

Note for payment: @Pujan92 is helping us. Although we don't have to pay $1,000 to @Pujan92, we should have partial payment. We can discuss the amount when the PR is done

**Comment by hayata-suenaga** [comment]
@Pujan92 @pradeepmdk 

Please confirm if this issue is related to this one: https://github.com/Expensify/App/issues/2653

**Comment by pradeepmdk** [comment]
@hayata-suenaga no

**Comment by Pujan92** [comment]
I am seeing other occurrences  of the same too, so passing `|| ''` all places might be isn't a good option.  
https://github.com/Expensify/App/blob/91ff78db32f0284a73db13b2e2372583d71cb27f/src/components/ReportActionItem/MoneyRequestView.js#L91
https://github.com/Expensify/App/blob/91ff78db32f0284a73db13b2e2372583d71cb27f/src/components/Attachments/AttachmentCarousel/extractAttachmentsFromReport.js#L54

My suggestion will be first we need to update the backend if we are considering updating `receiptFilename` to `filename` and then we can try to reproduce this issue.

**Comment by hayata-suenaga** [comment]
the internal team is working on the backend change now. will get back to you once the change is done 🙇

**Comment by pradeepmdk** [comment]
@Pujan92 @hayata-suenaga we need to fix this wrong source mapping Also we need to fix this right?

**Comment by hayata-suenaga** [comment]
@pradeepmdk 

> we need to fix this wrong source mapping 

which one are you taking about?

**Comment by Pujan92** [comment]
> @Pujan92 @hayata-suenaga we need to fix this wrong source mapping Also we need to fix this right?

Sorry but I am not sure about this. Bcoz I don't have an idea about how local image converted to number instead of string in android. I can check tomorrow and update if needed.

**Comment by pradeepmdk** [comment]
@Pujan92 when its not converting its referenced when you import both android and ios

@hayata-suenaga 
>which one are you taking about?

in the current PR is allowed to load a local image until get the filename.

**Comment by hayata-suenaga** [comment]
wait. I tired to reproduce this on staging but couldn't

https://github.com/Expensify/App/assets/98560306/c94aa2b1-842b-4f04-90e1-3b4d5d2e75ee

**Comment by pradeepmdk** [comment]
@hayata-suenaga As I said this crash will happen only on native devices specifically on Android because of `Image source`.
web you don't have any problem.

**Comment by hayata-suenaga** [comment]
ah yea haha my brain is not working right now 😭

**Comment by pradeepmdk** [comment]
@hayata-suenaga cool...

**Comment by hayata-suenaga** [comment]
I'm still trying to understand the issue

https://github.com/Expensify/App/blob/91ff78db32f0284a73db13b2e2372583d71cb27f/src/libs/ReceiptUtils.js#L40-L66

`filename` is not used to create a uri in this function. Why does it matter that the `filename` being `undefined`?

**Comment by pradeepmdk** [comment]
@hayata-suenaga thanks for joining... I think now we are clear. let me know if you need anything. 👍

**Comment by situchan** [comment]
@hayata-suenaga if filename is undefined, it crashes in first line - `Str.isImage`
As you see crash log I reported in slack - https://expensify.slack.com/archives/C049HHMV9SM/p1693763384079999

<img width="1196" alt="Screenshot 2023-09-03 at 9 48 22 PM" src="https://github.com/Expensify/App/assets/108292595/3d3819a3-a1c7-4308-adc1-fd101d9b296a">

**Comment by situchan** [comment]
```
    isImage(url) {
        return _.contains(['jpeg', 'jpg', 'gif', 'png', 'bmp', 'webp'], this.getExtension(url));
    },
```

```
    getExtension(url) {
        return _.first(_.last(url.split('.')).split('?')).toLowerCase();
    },
```

**Comment by hayata-suenaga** [comment]
@JmillsExpensify @situchan is also a reporter

<img width="500" alt="Screenshot 2023-09-03 at 8 53 26 PM" src="https://github.com/Expensify/App/assets/98560306/236c8960-c22e-4d91-a70e-3c3dc20d1d71">

**Comment by mountiny** [comment]
@pradeepmdk @Pujan92 @parasharrajat I cannot repro this crash in staging, can you please confirm as well before I check this off from the deployer list?

**Comment by mountiny** [comment]
QA confirmed this is good.

**Comment by JmillsExpensify** [comment]
@parasharrajat Do we need to add a regression test to prevent this from happening in the future?

**Comment by JmillsExpensify** [comment]
In the meantime, payment summary is as follows:

 - Issue reporter: @situchan $50
 - Contributor: @Pujan92 $1,500 (incl. urgency bonus)
 - Contributor: @pradeepmdk $1,500 (incl. urgency bonus)
 - Contributor+: @parasharrajat $1,500 (incl. urgency bonus; paid via NewDot)

@hayata-suenaga Can you confirm this summary is right, or are the contributors splitting the payment?

**Comment by hayata-suenaga** [comment]
https://github.com/Expensify/App/issues/26585#issuecomment-1704400588#27221

@pradeepmdk made the PR but @Pujan92 helped us along the way. I think the half $750 is appropriate for @Pujan92's help

**Comment by situchan** [comment]
@JmillsExpensify I C+ reviewed [PR](https://github.com/Expensify/App/pull/26611)

**Comment by JmillsExpensify** [comment]
Ok thanks everyone! I've updated the payment summary once more. It should now be accurate based on @hayata-suenaga's comment above. We're ready for payments.

**Comment by JmillsExpensify** [comment]
@situchan offer sent via upwork. I didn't include the issue reporting, though feel free to update or I will pay that one when I pay/close the contract. 

@pradeepmdk I've issued payment based on the summary. 

@Pujan92 I've also issued payment based on the summary.

**Comment by situchan** [comment]
@JmillsExpensify yes, I am also eligible for reporting based on https://github.com/Expensify/App/issues/26585#issuecomment-1704572774

**Comment by JmillsExpensify** [comment]
>@situchan offer sent via upwork. I didn't include the issue reporting, though feel free to update or I will pay that one when I pay/close the contract.

Yes agreed.

**Comment by JmillsExpensify** [comment]
@hayata-suenaga Can you please comment on @parasharrajat's post above? I'm not sure why we'd pay $1,500 for both of C+, just like we didn't do that for Contributor either.

**Comment by hayata-suenaga** [comment]
waiting for payment

**Comment by JmillsExpensify** [comment]
> In the meantime, payment summary is as follows:
> 
> * Issue reporter: @situchan $50
> * Contributor: @Pujan92 $750 (incl. urgency bonus)
> * Contributor: @pradeepmdk $1,500 (incl. urgency bonus)
> * Contributor+: @situchan $1,000 (incl. urgency bonus; paid via NewDot)
> 
> @hayata-suenaga Can you confirm this summary is right, or are the contributors splitting the payment?

Going with this based on comments above.

**Comment by JmillsExpensify** [comment]
@situchan paid for reporting and C+. @Pujan92 paid for contributor work. @pradeepmdk paid for contributor work. @parasharrajat is paid via NewDot. Now that the regression test is created, I think we're all done here. Closing.

**Comment by JmillsExpensify** [comment]
$1,500 payment approved for @parasharrajat based on comment above.
